src/core/backbone_router/ndproxy_table.hpp

```diff
@@ -260,17 +262,9 @@ class NdProxyTable : public InstanceLocator, private NonCopyable
         Iterator(Instance &aInstance, Filter aFilter);
         Iterator(Instance &aInstance, IteratorType);
 
-        bool           IsDone(void) const { return (mCurrent == nullptr); }
-        void           Advance(void);
-        void           operator++(void) { Advance(); }
-        void           operator++(int) { Advance(); }
-        const NdProxy &operator*(void)const { return *mCurrent; }
-        bool           operator==(const Iterator &aOther) const { return mCurrent == aOther.mCurrent; }
-        bool           operator!=(const Iterator &aOther) const { return !(*this == aOther); }
-        NdProxy *      operator->(void) { return mCurrent; }
-        NdProxy &      operator*(void) { return *mCurrent; }
-
-        NdProxy *mCurrent;
+        void Advance(void);
+
+        NdProxy *mItem;
```

Comment:
I'm afraid not. As I understand, this CRTP pattern has to be an 'interface' and cannot have any members. Note that:
```
bool IsDone(void) const { return (static_cast<const IteratorType *>(this)->mItem == nullptr); }
```
It `static_cast` itself to its derived class. It may go wrong if this base class has some members. Looking at `LinkedListEntry` and `Timer` in current code, the member `mNext` is also in `Timer` instead of `LinkedListEntry`. Thoughts?

Reply:
I think we can use CRTP model can have members as well. 

`static_cast` between a base class and sub-class should always be safe and work correctly,  no matter if the base class has member variable or not (or if subclass has other base classes).  In `static_cast` the compiler will actually correctly handle the pointer conversion and may change the raw pointer address value when casting from base to sub class (e.g., if the base and sub-class are not aligned in memory, i.e. don't start from same address). The only requirement when casting from base to sub class is to be sure that the pointer to base class we have is indeed coming from a subclass instance, and I think this would be always the case in CRTP inheritance. `reinterpret_cast` on other hand won't be correct and/or safe.

For `LinkedListEntry` the reason we could not define the `mNext` in this it, is that in some of the cases the type itself is part of public OT APIs (which is a C based and is a `struct`) and therefore already defines the `mNext` field (e.g. `otNetifAddress).

I think with this type, we intend to use it only in the core and for C++ types, so we can consider having it define the `mItem` itself. We probably need it to be `protected` so that the inheriting type can also access it (it probably needs to implement `Advance()`).